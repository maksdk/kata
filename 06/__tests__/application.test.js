import fs from 'fs';
import path from 'path';
import {
  html
} from 'js-beautify';
import timer from 'timer-promise';
import userEvent from '@testing-library/user-event';
import nock from 'nock';

import run from '../src/application';

nock.disableNetConnect();

const htmlOptions = {
  preserve_newlines: true,
  unformatted: [],
};

const fixturesPath = path.join(__dirname, '__fixtures__');
const getTree = () => html(document.body.innerHTML, htmlOptions);

beforeEach(() => {
  const initHtml = fs.readFileSync(path.join(fixturesPath, 'index.html')).toString();
  document.documentElement.innerHTML = initHtml;
  run();
  nock('http://localhost')
    .post('/users')
    .reply(200);
});

test('application', async () => {
  expect(getTree()).toMatchSnapshot();

  const form = document.querySelector('[data-form="sign-up"]');
  const nameInput = document.getElementById('sign-up-name');
  const emailInput = document.getElementById('sign-up-email');
  const passwordInput = document.getElementById('sign-up-password');
  const passwordConfirmationInput = document.getElementById('sign-up-password-confirmation');

  nameInput.setAttribute('value', 'Petya');
  await userEvent.type(nameInput, 'Petya', {
    allAtOnce: true
  });
  emailInput.setAttribute('value', 'wrong-email');
  await userEvent.type(emailInput, 'wrong-email', {
    allAtOnce: true
  });
  passwordInput.setAttribute('value', 'long password without confirmation');
  await userEvent.type(passwordInput, 'long password without confirmation', {
    allAtOnce: true
  });

  await timer.start(10);
  expect(getTree()).toMatchSnapshot();

  emailInput.setAttribute('value', '');
  await userEvent.type(emailInput, '', {
    allAtOnce: true
  });
  passwordInput.setAttribute('value', 'qwert');
  await userEvent.type(passwordInput, 'qwert', {
    allAtOnce: true
  });

  await timer.start(10);
  expect(getTree()).toMatchSnapshot();

  emailInput.setAttribute('value', 'support@hexlet.io');
  await userEvent.type(emailInput, 'support@hexlet.io', {
    allAtOnce: true
  });
  passwordConfirmationInput.setAttribute('value', 'qwerty');
  await userEvent.type(passwordConfirmationInput, 'qwerty', {
    allAtOnce: true
  });

  await timer.start(10);
  expect(getTree()).toMatchSnapshot();

  form.dispatchEvent(new Event('submit'));
  await timer.start(10);
  await timer.start(10);
  await timer.start(10);
  await timer.start(10);
  expect(getTree()).toMatchSnapshot();
});

test('application', async () => {
  expect(getTree()).toMatchSnapshot();
});