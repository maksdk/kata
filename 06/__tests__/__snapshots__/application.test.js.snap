// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application 1`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"row\\">
        <div class=\\"col-sm-8 col-12\\">
            <div class=\\"card\\">
                <div class=\\"card-body\\">
                    <h5 class=\\"card-title\\">Sign up</h5>
                    <div data-container=\\"sign-up\\">
                        <form data-form=\\"sign-up\\" method=\\"post\\">
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-name\\">Name</label>
                                <input id=\\"sign-up-name\\" type=\\"text\\" class=\\"form-control\\" name=\\"name\\">
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-email\\">Email<sup>*</sup></label>
                                <input id=\\"sign-up-email\\" required=\\"\\" type=\\"email\\" class=\\"form-control\\" name=\\"email\\">
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-password\\">Password<sup>*</sup></label>
                                <input id=\\"sign-up-password\\" required=\\"\\" type=\\"password\\" class=\\"form-control\\" name=\\"password\\">
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-password-confirmation\\">Password Confirmation<sup>*</sup></label>
                                <input id=\\"sign-up-password-confirmation\\" required=\\"\\" type=\\"password\\" class=\\"form-control\\" name=\\"passwordConfirmation\\">
                            </div>
                            <input type=\\"submit\\" class=\\"btn btn-primary\\" disabled=\\"\\" value=\\"Submit\\">
                        </form>
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>"
`;

exports[`application 2`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"row\\">
        <div class=\\"col-sm-8 col-12\\">
            <div class=\\"card\\">
                <div class=\\"card-body\\">
                    <h5 class=\\"card-title\\">Sign up</h5>
                    <div data-container=\\"sign-up\\">
                        <form data-form=\\"sign-up\\" method=\\"post\\">
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-name\\">Name</label>
                                <input id=\\"sign-up-name\\" type=\\"text\\" class=\\"form-control\\" name=\\"name\\" value=\\"Petya\\">
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-email\\">Email<sup>*</sup></label>
                                <input id=\\"sign-up-email\\" required=\\"\\" type=\\"email\\" class=\\"form-control is-invalid\\" name=\\"email\\" value=\\"wrong-email\\">
                                <div class=\\"invalid-feedback\\">Value is not a valid email</div>
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-password\\">Password<sup>*</sup></label>
                                <input id=\\"sign-up-password\\" required=\\"\\" type=\\"password\\" class=\\"form-control is-invalid\\" name=\\"password\\" value=\\"long password without confirmation\\">
                                <div class=\\"invalid-feedback\\">Password confirmation does not match to password</div>
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-password-confirmation\\">Password Confirmation<sup>*</sup></label>
                                <input id=\\"sign-up-password-confirmation\\" required=\\"\\" type=\\"password\\" class=\\"form-control\\" name=\\"passwordConfirmation\\">
                            </div>
                            <input type=\\"submit\\" class=\\"btn btn-primary\\" disabled=\\"\\" value=\\"Submit\\">
                        </form>
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>"
`;

exports[`application 3`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"row\\">
        <div class=\\"col-sm-8 col-12\\">
            <div class=\\"card\\">
                <div class=\\"card-body\\">
                    <h5 class=\\"card-title\\">Sign up</h5>
                    <div data-container=\\"sign-up\\">
                        <form data-form=\\"sign-up\\" method=\\"post\\">
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-name\\">Name</label>
                                <input id=\\"sign-up-name\\" type=\\"text\\" class=\\"form-control\\" name=\\"name\\" value=\\"Petya\\">
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-email\\">Email<sup>*</sup></label>
                                <input id=\\"sign-up-email\\" required=\\"\\" type=\\"email\\" class=\\"form-control\\" name=\\"email\\" value=\\"\\">
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-password\\">Password<sup>*</sup></label>
                                <input id=\\"sign-up-password\\" required=\\"\\" type=\\"password\\" class=\\"form-control is-invalid\\" name=\\"password\\" value=\\"qwert\\">
                                <div class=\\"invalid-feedback\\">Must be at least 6 letters</div>
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-password-confirmation\\">Password Confirmation<sup>*</sup></label>
                                <input id=\\"sign-up-password-confirmation\\" required=\\"\\" type=\\"password\\" class=\\"form-control\\" name=\\"passwordConfirmation\\">
                            </div>
                            <input type=\\"submit\\" class=\\"btn btn-primary\\" disabled=\\"\\" value=\\"Submit\\">
                        </form>
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>"
`;

exports[`application 4`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"row\\">
        <div class=\\"col-sm-8 col-12\\">
            <div class=\\"card\\">
                <div class=\\"card-body\\">
                    <h5 class=\\"card-title\\">Sign up</h5>
                    <div data-container=\\"sign-up\\">
                        <form data-form=\\"sign-up\\" method=\\"post\\">
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-name\\">Name</label>
                                <input id=\\"sign-up-name\\" type=\\"text\\" class=\\"form-control\\" name=\\"name\\" value=\\"Petya\\">
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-email\\">Email<sup>*</sup></label>
                                <input id=\\"sign-up-email\\" required=\\"\\" type=\\"email\\" class=\\"form-control\\" name=\\"email\\" value=\\"support@hexlet.io\\">
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-password\\">Password<sup>*</sup></label>
                                <input id=\\"sign-up-password\\" required=\\"\\" type=\\"password\\" class=\\"form-control is-invalid\\" name=\\"password\\" value=\\"qwert\\">
                                <div class=\\"invalid-feedback\\">Must be at least 6 letters</div>
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-password-confirmation\\">Password Confirmation<sup>*</sup></label>
                                <input id=\\"sign-up-password-confirmation\\" required=\\"\\" type=\\"password\\" class=\\"form-control\\" name=\\"passwordConfirmation\\" value=\\"qwerty\\">
                            </div>
                            <input type=\\"submit\\" class=\\"btn btn-primary\\" disabled=\\"\\" value=\\"Submit\\">
                        </form>
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>"
`;

exports[`application 5`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"row\\">
        <div class=\\"col-sm-8 col-12\\">
            <div class=\\"card\\">
                <div class=\\"card-body\\">
                    <h5 class=\\"card-title\\">Sign up</h5>
                    <div data-container=\\"sign-up\\">User Created!</div>
                </div>
            </div>
        </div>
    </div>
</div>"
`;

exports[`application 6`] = `
"<div class=\\"container mt-5\\">
    <div class=\\"row\\">
        <div class=\\"col-sm-8 col-12\\">
            <div class=\\"card\\">
                <div class=\\"card-body\\">
                    <h5 class=\\"card-title\\">Sign up</h5>
                    <div data-container=\\"sign-up\\">
                        <form data-form=\\"sign-up\\" method=\\"post\\">
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-name\\">Name</label>
                                <input id=\\"sign-up-name\\" type=\\"text\\" class=\\"form-control\\" name=\\"name\\">
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-email\\">Email<sup>*</sup></label>
                                <input id=\\"sign-up-email\\" required=\\"\\" type=\\"email\\" class=\\"form-control\\" name=\\"email\\">
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-password\\">Password<sup>*</sup></label>
                                <input id=\\"sign-up-password\\" required=\\"\\" type=\\"password\\" class=\\"form-control\\" name=\\"password\\">
                            </div>
                            <div class=\\"form-group\\">
                                <label for=\\"sign-up-password-confirmation\\">Password Confirmation<sup>*</sup></label>
                                <input id=\\"sign-up-password-confirmation\\" required=\\"\\" type=\\"password\\" class=\\"form-control\\" name=\\"passwordConfirmation\\">
                            </div>
                            <input type=\\"submit\\" class=\\"btn btn-primary\\" disabled=\\"\\" value=\\"Submit\\">
                        </form>
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>"
`;
